khook/x86/hook.c

```diff
@@ -1,22 +1,112 @@
 #include "../internal.h"
 
-#define kernel_write_enter() asm volatile (	\
+#define ASM_CLI asm volatile ("cli\n\t" ::: )
+#define ASM_STI asm volatile ("sti\n\t" ::: )
+
+#ifndef USE_CR0_FAM
+# ifndef USE_PTE_FAM
+#  error "At least one FAM *must* be used..."
+# endif
+#endif
+
+#ifdef USE_CR0_FAM
+# define kernel_write_enter() asm volatile (	\
 	"cli\n\t"				\
 	"mov %%cr0, %%rax\n\t"			\
 	"and $0xfffffffffffeffff, %%rax\n\t"	\
 	"mov %%rax, %%cr0\n\t"			\
 	::: "%rax" )
 
-#define kernel_write_leave() asm volatile (	\
+# define kernel_write_leave() asm volatile (	\
 	"mov %%cr0, %%rax\n\t"			\
 	"or $0x0000000000010000, %%rax\n\t"	\
 	"mov %%rax, %%cr0\n\t"			\
 	"sti\n\t"				\
 	::: "%rax" )
-
+#endif
+#ifdef USE_PTE_FAM
+/* Because of syncronization mechanism
+ * this set_rw/set_ro can be used only pairy
+ * and should be used in small time window
+ **/
+static int
+	is_first_rw = 0x00,
+	is_last_rw = 0x00;
+static volatile int lock_pte = 0x00;
+/*|SPIN FAMILY|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//**/
+/* argument MUST take address to global volatile int */
+/*|*/static void lock(volatile int *spin__) {
+/*|*/    while (__sync_lock_test_and_set(spin__, 1)) { ; }
+/*|*/}
+/*|*/static void unlock(volatile int *spin__) {
+/*|*/    __sync_synchronize(); /* Memory barrier. */
+/*|*/    *spin__ = 0x00;
+/*|*/}
+/*|PTE FAMILY|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*/
+/*|*/static void set_addr_rw(volatile uint64_t addr, int *flag){
+/*|*/   unsigned int level;
+/*|*/   uint64_t *pte = NULL;
+/*|*/   pte = (uint64_t*)lookup_address((unsigned long)addr, &level);
+/*|*/   if (pte[0x00] & ~((unsigned long long int)1<<1))
```

Comment:
It's better to use PTE flag name rather then just `1<<1`

Reply:
Do not use 0x.. prefix when it doesn't make sense

Reply:
Okey, i'll fix it